### PR TITLE
[serial] Serial Binding

### PR DESCRIPTION
**Pull request by MikeJMajor**
Implementation of a serial binding for OH 3.0. 

Provides similar functionality to the OH 1 binding.

Supports a bridge thing which represents a serial port and has a string channel for receiving/sending data and a trigger channel which fires when data is received. An image channel is provided to handle sending/receiving base64 encoded data.

Things have a filter and will only update their channels if the filter matches the received data. Things can be used to represent individual devices for serial data which is terminated with a line break for each device. The thing supports a single string data channel which can be linked to items with profiles used to transform the data.

E.g. A thing could be used to represent the following temperature sensor:

20;05;Cresta;ID=2801;TEMP=24;HUM=53;BAT=OK;

The filter would be set to filter on data starting with "20;05;Cresta;ID=2801;"

The data channel can be linked to a Number:Temperature item with a REGEX profile to extract the temperature, e.g. .*?TEMP=(.*?);.*
The data channel can be linked to a Number item with a REGEX profile to extract the humidity, e.g. .*?HUM=(.*?);.*

### CODE REVIEW

**Comment by Skinah** [comment]
Thanks for starting the binding. Just a heads up that you are using 
org.apache.commons.lang.StringUtils
You will not get the binding merged if you use any apache.commons libraries as they are getting removed from all bindings at the moment if not already. Need to find another way without those libs.

**Comment by MikeJMajor** [comment]
Code is ready for review. I'll work on the docs after the code has been reviewed.

**Review by J-N-K — CHANGES_REQUESTED** [review]
Also please add to the openhab-addons bom and CODEOWNERS.

Thanks for your contribution. I have left some comments in the code. If you have any wuestions, do not hesitate to ask.

File: bundles/org.openhab.binding.serial/src/main/java/org/openhab/binding/serial/internal/handler/SerialBridgeHandler.java
Comment (MikeJMajor) [reply bundles/org.openhab.binding.serial/src/main/java/org/openhab/binding/serial/internal/handler/SerialBridgeHandler.java:197]: Is the sleep that much of an issue?

I've refactored it and it appears to work, but now I'm concerned about 2 threads being involved and missing data. In my refactored code serialEvent checks if there's a ScheduledFuture active and does nothing. The assumption here is that receiveAndProcess will read in the data, but there's a risk that we are in the processing bit of receiveAndProcess and we won't do any more reads until we get another callback from serialEvent.

File: bundles/org.openhab.binding.serial/src/main/java/org/openhab/binding/serial/internal/handler/SerialBridgeConfiguration.java
Comment (MikeJMajor) [reply bundles/org.openhab.binding.serial/src/main/java/org/openhab/binding/serial/internal/handler/SerialBridgeConfiguration.java:44]: Are you suggesting changing parity to be an enum rather than a string?

How would you handle a user setting an invalid value in the config. I've tried it and you get a an IllegalArgumentException thrown from ConfigMapper.

File: bundles/org.openhab.binding.serial/src/main/java/org/openhab/binding/serial/internal/handler/SerialBridgeHandler.java
Comment (J-N-K) [reply bundles/org.openhab.binding.serial/src/main/java/org/openhab/binding/serial/internal/handler/SerialBridgeHandler.java:197]: You could create a `AtomicBoolean serialProcessing`, set it when you first call the read-method and unset it after you stopped reading. That makes sure it's never called twice. The problem with `Thread.sleep` is when you have (e.g. five) com-ports reading, it'll sleep all threads in the thread-pool.

File: bundles/org.openhab.binding.serial/README.md
Comment (MikeJMajor) [reply bundles/org.openhab.binding.serial/README.md:70]: I'm under the impression that you need to use system.rawbutton if you want to use profiles such as rawbutton-toggle-switch.

**Comment by MikeJMajor** [comment]
@J-N-K  I've addressed your review comments apart from the trigger one which I've left a comment. Thanks for the review.

File: bundles/org.openhab.binding.serial/src/main/java/org/openhab/binding/serial/internal/handler/SerialBridgeConfiguration.java
Comment (J-N-K) [reply bundles/org.openhab.binding.serial/src/main/java/org/openhab/binding/serial/internal/handler/SerialBridgeConfiguration.java:43]: But that's up to the user if he/she configures wrong values. Your code is much simpler and less error prone if you use the enum.

File: bundles/org.openhab.binding.serial/src/main/java/org/openhab/binding/serial/internal/channel/DeviceChannel.java
Comment (J-N-K) [inline bundles/org.openhab.binding.serial/src/main/java/org/openhab/binding/serial/internal/channel/DeviceChannel.java:71]:
```suggestion
    public Optional<String> transformData(final String data) {
        return transform.apply(data);
    }
```

(see below)

File: bundles/org.openhab.binding.serial/src/main/java/org/openhab/binding/serial/internal/handler/SerialDeviceHandler.java
Comment (J-N-K) [inline bundles/org.openhab.binding.serial/src/main/java/org/openhab/binding/serial/internal/handler/SerialDeviceHandler.java:180]:
```suggestion
        if (channel != null) {
            channel.transformData(data).ifPresent(value -> updateState(channelUID, new StringType(value));
        }
```

You have a null-check in l.169, so requiring non-null for the parameter should be ok. I think the same applies to `mapCommand`, please check that.

**Review by J-N-K — CHANGES_REQUESTED** [review]
Thanks for your fast response. I have left some more comments around the transformations. Already looks good otherwise.

**Comment by MikeJMajor** [comment]
@J-N-K I've addressed your latest set of review comments regarding the transformations. Have I covered all your comments?

File: bundles/org.openhab.binding.serial/src/main/resources/OH-INF/thing/thing-types.xml
Comment (J-N-K) [inline bundles/org.openhab.binding.serial/src/main/resources/OH-INF/thing/thing-types.xml:21]: can be removed if you provide no options

File: bundles/org.openhab.binding.serial/src/main/resources/OH-INF/thing/thing-types.xml
Comment (J-N-K) [inline bundles/org.openhab.binding.serial/src/main/resources/OH-INF/thing/thing-types.xml:95]: see above

**Review by J-N-K — CHANGES_REQUESTED** [review]
Sorry, some more comments. I think these will be my last ones.

File: bundles/org.openhab.binding.serial/src/main/java/org/openhab/binding/serial/internal/handler/SerialBridgeHandler.java
Comment (MikeJMajor) [reply bundles/org.openhab.binding.serial/src/main/java/org/openhab/binding/serial/internal/handler/SerialBridgeHandler.java:282]: I'm using the data field to have access to the last received data to handle a refresh command. If I don't have the data field I can ignore the refresh command?

Comment (J-N-K) [reply bundles/org.openhab.binding.serial/src/main/java/org/openhab/binding/serial/internal/handler/SerialBridgeHandler.java:282]: Good point. I would suggest to use the parameter nevertheless and store it in the field additionally. You can then use that to pass it when `REFRESH` is called.

This of course means that you must keep a `refresh` method in the ThingHandler below. I have edited my suggestions there.

File: bundles/org.openhab.binding.serial/src/main/java/org/openhab/binding/serial/internal/handler/SerialDeviceHandler.java
Comment (J-N-K) [inline bundles/org.openhab.binding.serial/src/main/java/org/openhab/binding/serial/internal/handler/SerialDeviceHandler.java:52]:
```suggestion
    private @Nullable String lastValue;
```

File: bundles/org.openhab.binding.serial/src/main/java/org/openhab/binding/serial/internal/handler/SerialDeviceHandler.java
Comment (J-N-K) [inline bundles/org.openhab.binding.serial/src/main/java/org/openhab/binding/serial/internal/handler/SerialDeviceHandler.java:65]:
```suggestion
        if (command instanceof RefreshType) {
            DeviceChannel channel = channels.get(channelUID);
            String lastValue = this,lastValue;
            if (channel != null) {
                refresh(channelUID, channel, lastValue);
            }
        } else {
```

File: bundles/org.openhab.binding.serial/src/main/java/org/openhab/binding/serial/internal/handler/SerialDeviceHandler.java
Comment (J-N-K) [inline bundles/org.openhab.binding.serial/src/main/java/org/openhab/binding/serial/internal/handler/SerialDeviceHandler.java:174]:
```suggestion
    private void refresh(ChannelUID channelUID, DeviceChannel channel, @Nullable String value) {
        if (isLinked(channelUID) && value != null) {
            channel.transformData(data).ifPresent(value -> updateState(channelUID, new StringType(value)));
        }
    }
```

File: bundles/org.openhab.binding.serial/src/main/java/org/openhab/binding/serial/internal/channel/DeviceChannel.java
Comment (MikeJMajor) [reply bundles/org.openhab.binding.serial/src/main/java/org/openhab/binding/serial/internal/channel/DeviceChannel.java:56]: For cases when there is no value the result of just passing in the optional gives:
`Mapped command is 'Optional.empty'` which I don't think is as clear as  `Mapped command is 'null'`

**Comment by MikeJMajor** [comment]
@J-N-K  I've addressed your latest set of review comments apart from the debug logging change which I've left a comment.

**Review by J-N-K — APPROVED** [review]
LGTM. Now we need a second review from someone else. Thanks.

File: bundles/org.openhab.binding.serial/src/main/java/org/openhab/binding/serial/internal/handler/SerialBridgeHandler.java
Comment (fwolter) [inline bundles/org.openhab.binding.serial/src/main/java/org/openhab/binding/serial/internal/handler/SerialBridgeHandler.java:73]: To have the same behavior across all platforms and configurations, I'd use a static charset here.
```suggestion
    private Charset charset = StandardCharsets.UTF_8;
```

File: bundles/org.openhab.binding.serial/src/main/java/org/openhab/binding/serial/internal/handler/SerialDeviceHandler.java
Comment (fwolter) [inline bundles/org.openhab.binding.serial/src/main/java/org/openhab/binding/serial/internal/handler/SerialDeviceHandler.java:117]: Why is the bridge status updated when the thing has been initialized?

File: bundles/org.openhab.binding.serial/src/main/java/org/openhab/binding/serial/internal/handler/SerialDeviceHandler.java
Comment (fwolter) [inline bundles/org.openhab.binding.serial/src/main/java/org/openhab/binding/serial/internal/handler/SerialDeviceHandler.java:153]: The `BaseThingHandler` sets the Thing states accordingly when the bridge status changes. Is this method necessary?

File: bundles/org.openhab.binding.serial/src/main/resources/OH-INF/thing/thing-types.xml
Comment (fwolter) [inline bundles/org.openhab.binding.serial/src/main/resources/OH-INF/thing/thing-types.xml:56]: I think the brackets are not necessary.
```suggestion
					<option value="N">None</option>
```

**Review by fwolter — CHANGES_REQUESTED** [review]

File: bundles/org.openhab.binding.serial/src/main/resources/OH-INF/thing/thing-types.xml
Comment (J-N-K) [reply bundles/org.openhab.binding.serial/src/main/resources/OH-INF/thing/thing-types.xml:56]: I thought about that, too. But: Usually the settings are given as `8N1` or `7E0`, only one letter. And it's probably more intuitive for less experienced users to have it with the brackets.

File: bundles/org.openhab.binding.serial/src/main/java/org/openhab/binding/serial/internal/handler/SerialDeviceHandler.java
Comment (J-N-K) [reply bundles/org.openhab.binding.serial/src/main/java/org/openhab/binding/serial/internal/handler/SerialDeviceHandler.java:117]: In fact this sets the thing status depending on the bridge status by calling `bridgeStatusChanged` with the current bridge status. The other option would be to duplicate the code from `bridgeStatusChanged` here.

File: bundles/org.openhab.binding.serial/README.md
Comment (MikeJMajor) [reply bundles/org.openhab.binding.serial/README.md:70]: This was to provide some sort of replacement for the v1 "switch" channel which would just toggle ON/OFF when data was received.

File: bundles/org.openhab.binding.serial/src/main/java/org/openhab/binding/serial/internal/handler/SerialDeviceHandler.java
Comment (MikeJMajor) [reply bundles/org.openhab.binding.serial/src/main/java/org/openhab/binding/serial/internal/handler/SerialDeviceHandler.java:153]: This is preventing the thing from going online if there is a thing level configuration error.

File: bundles/org.openhab.binding.serial/README.md
Comment (fwolter) [reply bundles/org.openhab.binding.serial/README.md:70]: Is there any use case for it? Rules can be triggered event-based on Item/Channel state changes.

File: bundles/org.openhab.binding.serial/src/main/java/org/openhab/binding/serial/internal/handler/SerialDeviceHandler.java
Comment (fwolter) [reply bundles/org.openhab.binding.serial/src/main/java/org/openhab/binding/serial/internal/handler/SerialDeviceHandler.java:153]: I understood that the Thing should be online when there's no configuration error and the Bridge is online.

This behavior can be achieved by using the framework. You can set the Thing status to ONLINE when there is no configuration error. Then, the framework takes care, that the Thing is only online when the Bridge is.

Comment (MikeJMajor) [reply bundles/org.openhab.binding.serial/src/main/java/org/openhab/binding/serial/internal/handler/SerialDeviceHandler.java:153]: It seems the framework only takes care of it if you explicitly check the bridge status and don't put the thing online if the bridge is offline. The framework allows you to set the thing online even though the bridge is offline. I'll make that change since its cleaner not overriding `bridgeStatusChanged`.

Comment (fwolter) [reply bundles/org.openhab.binding.serial/src/main/java/org/openhab/binding/serial/internal/handler/SerialDeviceHandler.java:153]: From my knowledge the Thing doesn't go online when the bridge is offline, although if the binding sets it explicitly to online. But I'm open for conviction.

File: bundles/org.openhab.binding.serial/README.md
Comment (MikeJMajor) [reply bundles/org.openhab.binding.serial/README.md:70]: Shall I remove that channel since it just duplicates what the SerialButton binding does?

**Comment by trpkosj** [comment]
Hello Mike,

I hope it is not inapropriate to ask here about usage of this bindings... First of all thank you very much for development, I am using current version and in general it works very well.

I need to solve one configuration issue I do not uderstand. I have a roller shutter with possibility to lean which accepts these commands:
UP <5>
STOP <6>
DOWN <1>
DOWN and lean 1/3 <2>
DOWN and lean 2/3 <3>
DOWN and lean 3/3 <4>

UP, STOP and DOWN work well. But how can I configure those commands DOWN and lean?

I tried Percent Format in channel configuration. It works only for one such a command (or I do something worng). If this is the right way, how exactly shall I configure? I tried many ways, so just for one example 30='<2>;60='<3>;90='<4>'

Please do you have any advise? 
Also is it possible to use expressions in Percent Format (like = (items.x.state===30) ? '<4> : '<3>')

Thank you very much,
jiri

**Comment by MikeJMajor** [comment]
@trpkosj What controls are you trying to use to send those down and lean commands?

**Comment by trpkosj** [comment]
> 
> 
> @trpkosj What controls are you trying to use to send those down and lean commands?

I use OH 3 new UI, Label Card with Action: Send command to the rollershutter item with Action Command: 90

actually those commands are a bit more complicated, thay contain also address. It is exactly <Z21x> where Z21 is address and x is command (up, down, ...). There are many roller shutters listening on serial line.... so I need to have configuration somewhere in the channel. One channel is one rollershutter and one serialDevice has many channels....

**Comment by MikeJMajor** [comment]
@trpkosj so your commands are `UP, STOP, DOWN, 30, 60, 90` and you need to send something like `<address>:<command>`?

You could use some cascading regex expressions to handle the percentage values, e.g.

Screenshot_20201116-213126

Do you need to send 30, 60 and 90? You could avoid the need for the regex by sending 2, 3 and 4.

**Comment by trpkosj** [comment]
> 
> 
> @trpkosj so your commands are `UP, STOP, DOWN, 30, 60, 90` and you need to send something like `<address>:<command>`?
> 
> You could use some cascading regex expressions to handle the percentage values, e.g.
> 
> Screenshot_20201116-214258
> 
> Do you need to send 30, 60 and 90? You could avoid the need for the regex by sending 2, 3 and 4.

wow, that works, thank you very much, you saved me couple of days:-)

File: bundles/org.openhab.binding.serial/README.md
Comment (fwolter) [reply bundles/org.openhab.binding.serial/README.md:70]: If I see correctly, the Serial Button binding triggers when DTR or so is raised by a switch/contact and your Channel triggers when data is available. As the Channel's functionality is redundant to the event based framework, I'd remove it.

Comment (MikeJMajor) [reply bundles/org.openhab.binding.serial/README.md:70]: Unless I've missed something it's behaving exactly as the SerialButton binding, i.e. triggering based on the `SerialPortEvent.DATA_AVAILABLE` event.

So it's redundant in the sense that there is already a binding to provide that functionality, but it means some users may have to replace the use of one binding with two.

Comment (fwolter) [reply bundles/org.openhab.binding.serial/README.md:70]: If I see correctly, the Serial Button binding triggers also when RI/CTS is raised, not (only) when data is available. I think that's a difference to your binding and not a redundancy.

My point is that OH already provides the ability to react on a state change of a Channel/Item via rules. So, it's not necessary to have a dedicated Switch that changes when a Channel has been updated. Or do I miss something?

Comment (MikeJMajor) [reply bundles/org.openhab.binding.serial/README.md:70]: There's a simple use case of using the trigger channel with the toggle profile to hook it up to a switch without the need of a rule.

Comment (fwolter) [reply bundles/org.openhab.binding.serial/README.md:70]: I wasn't aware of that profile. You convinced me.

File: bundles/org.openhab.binding.serial/src/main/resources/OH-INF/thing/thing-types.xml
Comment (MikeJMajor) [reply bundles/org.openhab.binding.serial/src/main/resources/OH-INF/thing/thing-types.xml:56]: The use of brackets was copied from the DSMR binding.  Do you want me to keep the brackets or remove them?

Comment (fwolter) [reply bundles/org.openhab.binding.serial/src/main/resources/OH-INF/thing/thing-types.xml:56]: It's not that important. You can keep them.

**Comment by MikeJMajor** [comment]
@fwolter I believe I've addressed your remaining review comments. Is there anything outstanding?

**Review by fwolter — APPROVED** [review]
LGTM

**Pull request merged**
